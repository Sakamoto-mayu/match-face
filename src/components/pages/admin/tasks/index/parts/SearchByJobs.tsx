"use client";
import Input from "@/components/ui/Input";
import OrangeButton from "@/components/ui/button/OrangeButton";
import WhiteCheckButton from "@/components/ui/button/WhiteCheckButton";
import { SyntheticEvent, useMemo, useState } from "react";

// 削除予定
import { departments } from "@/const/tasks";
import Link from "next/link";
import { useRouter } from "next/navigation";

const SearchByJobs = () => {
  const [jobsFilter, setJobsFilter] = useState<string[]>([]); // 職種フィルター状態管理
  const [inputVal, setInputVal] = useState<string>(""); // 検索入力値状態管理

  const router = useRouter();

  // 検索条件をクエリに設定
  const query = useMemo(() => {
    if (jobsFilter.length === 0 && inputVal.length === 0) return "";

    const searchKeyword =
      inputVal.length !== 0
        ? `searchKeyword=${inputVal // ex)"a b c" → "a_b_c"
            .split(/[\s]+/)
            .filter((word) => word.length !== 0)
            .join("_")}`
        : "";

    const departmentId =
      jobsFilter.length !== 0
        ? `${searchKeyword && "&"}departmentId=${jobsFilter.join("_")}`
        : "";

    const searchQuery = `?${searchKeyword && searchKeyword}${departmentId}`;
    return searchQuery;
  }, [jobsFilter, inputVal]);

  // 検索入力値を適用
  const handleSearch = (e: SyntheticEvent) => {
    e.preventDefault();
    router.push(`/admin/tasks${query}`);
  };

  const handleSetFilter = (department: string) => {
    // 職種レコードを挿入
    if (!jobsFilter.includes(department)) {
      const newArr = [...jobsFilter];
      newArr.push(department);
      setJobsFilter(newArr);
    } else {
      // 職種レコードを削除
      const newArr = jobsFilter.filter((job) => {
        return job !== department;
      });
      setJobsFilter(newArr);
    }
    // console.log("jobsFilter2", jobsFilter);
  };

  // 検索条件を含めたクエリを生成

  return (
    <div className="border-2 rounded-md w-3/6 mx-auto mt-6 p-3">
      {/* 職種フィルター入力欄 */}
      <div className="w-full mb-4">
        <div className="mb-2">
          <h2 className="text-base ml-3">▶️募集職種で検索</h2>
        </div>
        <div className="flex justify-center">
          {departments.map((department) => {
            return (
              <WhiteCheckButton
                key={department.id}
                label={department.name}
                className="text-xs w-16 mx-3"
                onClick={() => handleSetFilter(department.name)}
              />
            );
          })}
        </div>
      </div>

      {/* 案件名/ユーザー名検索欄 */}
      <div className="w-full">
        <div className="mb-2">
          <h2 className="text-base ml-3">▶️案件名/ユーザー名で検索</h2>
        </div>
        <form onSubmit={handleSearch} className="flex flex-col items-center">
          <Input
            id="search"
            className="border-light-gray text-xs p-1 w-[35rem]"
            onChange={(e) => setInputVal(e.target.value)}
            data-testid="search-box"
          />
          <OrangeButton
            label="検索"
            className="w-20 text-sm mt-2"
            type="submit"
          />
        </form>
      </div>
    </div>
  );
};

export default SearchByJobs;
