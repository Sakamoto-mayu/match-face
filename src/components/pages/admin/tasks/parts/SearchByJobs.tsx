import Input from "@/components/ui/Input";
import WhiteButton from "@/components/ui/button/WhiteButton";

const SearchByJobs = () => {
  return (
    <div>
      <div className="flex items-center">
        <Input id="search"/>
        <WhiteButton label="検索" textSize="text-sm"/>
      </div>
      
    </div>
  );
};

export default SearchByJobs;
