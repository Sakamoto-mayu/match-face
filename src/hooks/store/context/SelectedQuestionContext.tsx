"use client";
import {
  Dispatch,
  ReactNode,
  Reducer,
  createContext,
  useContext,
  useReducer,
} from "react";
import { selectedQuestionReducer } from "../reducer/selectedQuestionReducer";
import { SelectReducerAction } from "@/types/admin/tasks/register/types";

const SelectedQuestionContext = createContext<
  [string[], Dispatch<SelectReducerAction>] | undefined
>(undefined);

export const SelectedQuestionProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer<Reducer<string[], SelectReducerAction>>(
    selectedQuestionReducer,
    []
  );
  return (
    <SelectedQuestionContext.Provider value={[state, dispatch]}>
      {children}
    </SelectedQuestionContext.Provider>
  );
};

export const useSelectQuestion = () => {
  const context = useContext(SelectedQuestionContext);
  if (context === undefined)
    throw new Error("Function 'useUserSelect' must be used in Provider.");
  return context;
};
