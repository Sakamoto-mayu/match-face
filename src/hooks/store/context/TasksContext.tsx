"use client";

import React from "react";

// 職種フィルター用context
const JobsFilterContext = React.createContext<
  [string[], React.Dispatch<React.SetStateAction<string[]>>] | undefined
>(undefined);

export const JobsFilterProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const [filterlist, setFilterList] = React.useState<string[]>([]);
  return (
    <JobsFilterContext.Provider value={[filterlist, setFilterList]}>
      {children}
    </JobsFilterContext.Provider>
  );
};

export const useJobsFilter = () => {
  const context = React.useContext(JobsFilterContext);
  if (context === undefined) {
    throw new Error("useRefine must be used within a Provider");
  }
  return context;
};
// ここまで
