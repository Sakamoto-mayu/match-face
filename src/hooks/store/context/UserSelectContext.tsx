"use client";
import {
  Dispatch,
  ReactNode,
  Reducer,
  createContext,
  useContext,
  useReducer,
} from "react";
import { userSelectReducer } from "../reducer/userSelectReducer";
import { UserSelectReducerAction } from "@/types/admin/tasks/register/types";

const UserSelectContext = createContext<
  [string[], Dispatch<UserSelectReducerAction>] | undefined
>(undefined);

export const CounterProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer<Reducer<string[], UserSelectReducerAction>>(
    userSelectReducer,
    []
  );
  return (
    <UserSelectContext.Provider value={[state, dispatch]}>
      {children}
    </UserSelectContext.Provider>
  );
};

export const useUserSelect = () => {
  const context = useContext(UserSelectContext);
  if (context === undefined)
    throw new Error("Function 'useCounter' must be used in Provider.");
  return context;
};
