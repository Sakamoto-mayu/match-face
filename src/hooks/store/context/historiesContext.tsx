'use client'
import { createContext, useContext } from "react";
import { Departments, Skills } from "@/types/admin/histories/admin_histories";
import { skills,departments } from "@/const/admin_histories";

export const SkillsContext = createContext<Skills|undefined>(undefined)
export const DepartmentsContext = createContext<Departments|undefined>(undefined)

// constからのimport(valueの値)をfetchで取得したデータに変える
// const skills = fetch('')
// const departments = fetch('')

export function HistoriesProvider({children}:{children:React.ReactElement}) {
    return (
        <SkillsContext.Provider value = {skills}>
        <DepartmentsContext.Provider value = {departments}>
            {children}
        </DepartmentsContext.Provider>
        </SkillsContext.Provider>
    )
}
export const useSkills = () => {
    const context = useContext(SkillsContext)
    if(context === undefined){
        throw new Error('skillsが見つかりません')
    }
    return context
}
export const useDepartments = () => {
    const context = useContext(DepartmentsContext)
    if(context === undefined){
        throw new Error('departmentsが見つかりません')
    }
    return context
}
