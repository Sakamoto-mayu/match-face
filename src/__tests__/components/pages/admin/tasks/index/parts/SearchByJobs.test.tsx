import SearchByJobs from "@/components/pages/admin/tasks/index/parts/SearchByJobs";
import { FilterProvider } from "@/hooks/store/context/TasksContext";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";

describe("TaskList.tsx", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it("レンダリング時", () => {
    const { container } = render(
      <FilterProvider>
        <SearchByJobs />
      </FilterProvider>
    );
    expect(container).toMatchSnapshot();
  });
  describe("入力テスト", () => {
    beforeAll(() => {
      cleanup();
      render(
        <FilterProvider>
          <SearchByJobs />
        </FilterProvider>
      );
    });
    it("検索ボックス入力テスト", () => {
      const input = screen.getByTestId("search-box") as HTMLInputElement;
      fireEvent.change(input, { target: { value: "hoge" } });
      // 正常に入力されているかテスト
      expect(input.value).toBe("hoge");
    });
  });
});
