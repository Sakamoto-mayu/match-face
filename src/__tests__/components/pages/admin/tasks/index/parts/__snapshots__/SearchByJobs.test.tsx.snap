// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList.tsx レンダリング時 1`] = `
<div>
  <div
    class="flex flex-col items-center border-2 rounded-md w-3/6 mx-auto mt-6 p-3"
  >
    <div
      class="flex items-center mb-4"
    >
      <div>
        <input
          class="border border-light-gray text-xs p-1 w-96"
          id="search"
        />
      </div>
      <button
        class="bg-white hover:bg-gray-100 active:bg-gray-200 active:translate-y-0.5 active:shadow-sm py-2 px-4 rounded-full border shadow-md text-xs ml-2"
      >
        検索
      </button>
    </div>
    <div
      class="flex items-center mb-4"
    >
      <div
        class="mx-3"
      >
        <button
          class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
        >
          Java
        </button>
      </div>
      <div
        class="mx-3"
      >
        <button
          class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
        >
          PHP
        </button>
      </div>
      <div
        class="mx-3"
      >
        <button
          class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
        >
          FR
        </button>
      </div>
      <div
        class="mx-3"
      >
        <button
          class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
        >
          CL
        </button>
      </div>
      <div
        class="mx-3"
      >
        <button
          class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
        >
          ML
        </button>
      </div>
      <div
        class="mx-3"
      >
        <button
          class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md text-xs w-16"
        >
          QA
        </button>
      </div>
    </div>
    <div>
      <div>
        <button
          class="bg-orange drop-shadow-md hover:saturate-150 active:drop-shadow-none active:shadow-inner hover:bg-depp-orange active:bg-depp-orange active:translate-y-0.5 text-white rounded-xl p-2 h-auto w-28 text-sm"
        >
          絞り込み
        </button>
      </div>
    </div>
  </div>
</div>
`;
